src/sentry/static/sentry/app/utils/queryString.tsx

```diff
@@ -16,41 +17,127 @@ export function addQueryParamsToExistingUrl(
-
 /**
- * Append a tag key:value to a query string.
+ * Add a key-value to the queryString.
+ *
+ * If there is an existing value on the same key, this function will replace it.
+ *
+ * However, if the key is `query`, it will append the new value to the existing
+ * value, delimiting the values with a space.
  *
  * Handles spacing and quoting if necessary.
  */
-export function appendTagCondition(
-  query: QueryValue,
+export function addKeyValueToQueryString(
+  query: Query = {},
   key: string,
   value: string
-): string {
-  let currentQuery = Array.isArray(query) ? query.pop() : isString(query) ? query : '';
+): Query {
+  if (key === 'query') {
+    return addKeyValueToQueryStringQuery(query, value);
+  }
 
-  if (isString(value) && value.indexOf(' ') > -1) {
-    value = `"${value}"`;
+  // Wrap value if there are spaces
+  return {
+    ...query,
+    [key]: value.indexOf(' ') > -1 ? `"${value}"` : value,
+  };
+}
+
+/**
+ * Used by `addKeyValueToQueryStringQuery`, `decodeQueryValueToObject` and
+ * `encodeObjectToQueryValue` as a marker when they receive a value with either
+ * no keys or the wrong format.
+ */
+const NO_KEY = '__NO_KEY__';
+
+/**
+ * Append a tag (key:value) to QueryString.query. This method ensures uniqueness
+ * of keys in QueryString.query. If there is an existing key with a value, it
+ * will be overwritten by the new value.
+ *
+ * Handles spacing and quoting if necessary.
+ */
+export function addKeyValueToQueryStringQuery(qs: Query = {}, tag: string): Query {
+  const {query} = qs;
+  const queryObject = decodeQueryValueToObject(query);
+
+  let key, value;
+  const keyValue = tag.split(':');
+  if (keyValue.length === 2) {
+    [key, value] = keyValue;
+  } else {
+    key = NO_KEY;
+    value = tag;
   }
-  if (currentQuery) {
-    currentQuery += ` ${key}:${value}`;
+
+  // Wrap value if there are spaces
+  if (value.indexOf(' ') > -1) {
+    value = `"${value.trim()}"`;
+  }
+
+  // Add new key-value into QueryObject
+  if (key === NO_KEY) {
+    queryObject[key] = `${queryObject[key]} ${value}`.trim();
   } else {
-    currentQuery = `${key}:${value}`;
+    queryObject[key] = value.trim();
   }
 
-  return currentQuery;
+  return {
+    ...qs,
+    query: encodeObjectToQueryValue(queryObject),
+  };
+}
+
+/**
+ * This method specifically parses "?query=key:value" into an object and
+ * resolves duplicate keys. In the case of duplicate keys, the last value will
+ * be used.
+ */
+function decodeQueryValueToObject(query: Query['query']): {[key: string]: string} {
+  const currQuery: string = isString(query)
+    ? query
+    : Array.isArray(query)
+    ? query.reduce((acc, q) => `${acc} ${q}`, '')
+    : '';
+
+  // 1) Start with the QueryValue: "key1:value1 key2:value2"
+  // 2) Split it into their pairs: ["key1:value1", "key2:value2"]
+  // 3) Split it into key-values:  { key1: value1, key2: value2 }
+  return currQuery.split(' ').reduce((acc, keyValue) => {
+    const pair = keyValue.split(':');
```

Comment:
Quoted values could also contain `:`. For example `mytag:"namespace:value"`

Reply:
Would be nice to have a query tokenizer on the frontend